src/hotspot/share/prims/jvmtiImpl.cpp

```diff
@@ -365,27 +266,14 @@ int JvmtiBreakpoints::clear(JvmtiBreakpoint& bp) {
 }
 
 void JvmtiBreakpoints::clearall_in_class_at_safepoint(Klass* klass) {
-  bool changed = true;
-  // We are going to run thru the list of bkpts
-  // and delete some.  This deletion probably alters
-  // the list in some implementation defined way such
-  // that when we delete entry i, the next entry might
-  // no longer be at i+1.  To be safe, each time we delete
-  // an entry, we'll just start again from the beginning.
-  // We'll stop when we make a pass thru the whole list without
-  // deleting anything.
-  while (changed) {
-    int len = _bps.length();
-    changed = false;
-    for (int i = 0; i < len; i++) {
-      JvmtiBreakpoint& bp = _bps.at(i);
-      if (bp.method()->method_holder() == klass) {
-        bp.clear();
-        _bps.remove(i);
-        // This changed 'i' so we have to start over.
-        changed = true;
-        break;
-      }
+  assert(SafepointSynchronize::is_at_safepoint(), "must be at safepoint");
+
+  // Go backwards because this removes entries that are freed.
+  for (int i = length() - 1; i >= 0; i--) {
+    JvmtiBreakpoint& bp = at(i);
+    if (bp.method()->method_holder() == klass) {
+      bp.clear();
+      remove(i);
```

Comment:
After reading the original comments, I get the feeling that the concern was that the array might be re-organized (recached). The comment "...the next entry might no longer be at i+1" of course is true. It would be at i if there was no recaching, in which case you could use while loop starting at 0 and not increment `i` if the entry is deleted.